polkadot/xcm/pallet-xcm/src/lib.rs

```diff
@@ -1211,31 +1257,98 @@ impl<T: Config> Pallet<T> {
-		let max_assets = assets.len() as u32;
-		let assets: MultiAssets = assets.into();
-		let xcm = Xcm(vec![
-			BuyExecution { fees, weight_limit },
-			DepositAsset { assets: Wild(AllCounted(max_assets)), beneficiary },
-		]);
-		let mut message = Xcm(vec![
-			SetFeesMode { jit_withdraw: true },
-			TransferReserveAsset { assets, dest, xcm },
-		]);
-		let weight =
-			T::Weigher::weight(&mut message).map_err(|()| Error::<T>::UnweighableMessage)?;
-		let hash = message.using_encoded(sp_io::hashing::blake2_256);
-		let outcome =
-			T::XcmExecutor::execute_xcm_in_credit(origin_location, message, hash, weight, weight);
-		Self::deposit_event(Event::Attempted { outcome });
-		Ok(())
+		let (origin_location, mut assets) = value;
+
+		if fee_asset_item as usize >= assets.len() {
+			return Err(Error::<T>::Empty.into())
+		}
+		let mut fees = assets.swap_remove(fee_asset_item as usize);
+		let fees_transfer_type =
+			<T::XcmExecutor as AssetTransferSupport>::determine_for(&fees, &dest)
+				.map_err(Error::<T>::InvalidAsset)?;
+		let assets_transfer_type = if assets.is_empty() {
+			// Single asset to transfer (one used for fees where transfer type is determined above).
+			ensure!(fees_transfer_type != TransferType::Teleport, Error::<T>::Filtered);
+			fees_transfer_type
+		} else {
+			// Find reserve for non-fee assets.
+			Self::validate_assets_and_find_reserve(&assets, &dest)?
+		};
+
+		// Disallow (for now) different _remote_ reserves for assets and fees.
+		match (&fees_transfer_type, &assets_transfer_type) {
+			(TransferType::RemoteReserve(a), TransferType::RemoteReserve(b)) if a != b =>
+				return Err(Error::<T>::TooManyReserves.into()),
+			_ => (),
+		};
+
+		if fees_transfer_type == assets_transfer_type {
+			// Same reserve location (fees not teleportable), we can batch together fees and assets
+			// in same reserve-based-transfer.
+			assets.push(fees.clone());
+		} else {
+			// Different transfer types: we have to do separate transfers for fees and assets.
+			// This code block handles transferring the assets *used for fees*.
+
+			// Use only half the weight limit for fees transfer, then other half for assets
+			// transfer.
+			weight_limit = Self::halve_weight_limit(&weight_limit);
+
+			// When assets reserve is remote chain, we need to "prefund" fees to be able to
+			// BuyExecution on both chains. Split fees, and deposit half at assets-reserve chain
+			// and half at destination.
+			if let TransferType::RemoteReserve(assets_reserve) = assets_transfer_type {
+				// Halve amount of fees, each half will be sent to one chain.
+				Self::halve_fees(&mut fees)?;
+				// Halve weight limit again to be used for the two fees transfers.
+				let quarter_weight_limit = Self::halve_weight_limit(&weight_limit);
+				// Send half the `fees` to `beneficiary` on assets-reserve chain.
+				Self::build_and_execute_xcm_transfer_type(
+					origin_location,
+					assets_reserve,
+					beneficiary,
```

Comment:
I'm a bit confused here (trying to wrap my head around all this stuff :) ) - could you, please, verify + correct my thoughts: `beneficiary` is valid in the context of `dest` chain - e.g. `dest = ParentThen(X1(Parachain(42)))` and `beneficiary` is some account within this chain (`MultiLocation::new(0, X1(AccountId32([7u8; 32])))`) - so we are planning to move some funds to that account on sibling parachain `42`?

This call (`build_and_execute_xcm_transfer_type(dest = assets_reserve, beneficiary)`) will lead to depositing fees to the same account id, but on some other parachain if e.g. `assets_reserve = ParentThen(X1(Parachain(43)))`. Is that correct?

Reply:
Also couple of questions here:
1) am I right that we are going to deposit funds to the `beneficiary` account on a chain where it may not exist? Then what if `fees / 2` is less than the existential deposit? Is it the user responsibility to check that before?
2) also - what's the assumed state after all three programs have been computed on all involved chains. IIUC first two programs will deposit required `fees / 2` to *some* accounts on two chains. But after the 3rd program is executed AND the actual fee is less than the `fee / 2`, then is it possible that we leave remaining fee on some 3rd party chain? Just wanted to clarify

Reply:
> am I right that we are going to deposit funds to the beneficiary account on a chain where it may not exist? Then what if fees / 2 is less than the existential deposit? Is it the user responsibility to check that before?

yes, you are right - it would be the user responsibility to check it..

We could also add a config type that the runtime populates with a list/matrix of known/hrmp-connected chains and their known EDs (for at least the fees assets) - I'll try it out see how it looks, but that would require that we maintain these lists in runtimes which might not be desired? maybe make it optional? @joepetrowski WDYT?

>  then is it possible that we leave remaining fee on some 3rd party chain? Just wanted to clarify

yes, exactly - we do dumb split in half and hope each half covers costs on respective chain - unspent portion just remains there in `beneficiary` account on intermediate/remote chain (hey, at least it's not lost 😆 )

don't have any idea on how to make this better; seems like most of the fees correct usage and weight_limits correct usage is left in the user's responsibility for most things..

Reply:
> We could also add a config type that the runtime populates with a list/matrix of known/hrmp-connected chains and their known EDs (for at least the fees assets) - I'll try it out see how it looks, but that would require that we maintain these lists in runtimes which might not be desired? maybe make it optional? @joepetrowski WDYT?

Ok, thanks. IMO better to update call docs that to have that additional maintenance burden :) Especially since those `fees / 2`, `weight / 4` are very well hidden somewhere deep in the code. Maybe just describe this worst (in terms of involved networks) possible case + that fees + weight must be double checked.

But then there's probably another issue with that. Imagine that the ED on that 3rd (reserve) chain is `100`. Let's say `beneficiary` account exists there and has exactly `100` units. What if someone would send some minimal amount to `beneficiary` at `dest` chain and would select `fees` in such a way, so that after paying those fees the balance would be `99`. I.e. `fees / 2` won't be enough to cover all costs, but since `beneficiary` exists there and has a balance, the XCM would succeed, but it'd result in account death. Then someone could e.g. replay immortal transactions, submitted by that account. Do you think it is possible?

> yes, exactly - we do dumb split in half and hope each half covers costs on respective chain - unspent portion just remains there in beneficiary account on intermediate/remote chain (hey, at least it's not lost 😆 )

Thanks for clarifying. And then, going back to my first question. `beneficiary` is defined relative to `dest` - at least call docs say that. So what if `beneficiary` is not an account (because if you have a private key, then you indirectly "own" beneficiary with same public on other chains), but e.g. pallet or a contract. Then at `assets_reserve` it would be absolutely different entity - e.g. if it is the `X1(PalletInstance(42))`, then on `dest` it is the one pallet (actual `beneficiary`, specified by the caller), say, "Treasury" and at the `asset_chain` it could be e.g. "Grandpa" pallet. Is that possible? If so, then maybe we need to reanchor `dest` here, so that it'd point to the actual `beneficiary` location? Does it make sense?

Sorry - I may be totally wrong here, just trying to understand :)

Reply:
@svyatonik I've reworked the logic based on https://github.com/paritytech/polkadot-sdk/pull/1672#issuecomment-1775033549 and most of above went away. I will close this thread, so please have a look and open new threads for still open questions.

Reply:
Yep, I suppose it should be fixed now. IIUC we now send messages to `dest` only, so `beneficiary` is always treated within the assumed context. Re this ED part - I was not sure for previous case, but now I think everything is also good given supported cases.